src/client/Microsoft.Identity.Client/ManagedIdentity/V2/CertificateCache.cs

```diff
@@ -0,0 +1,226 @@
+        /// the next <see cref="GetAsync"/> will refresh it.
+        /// </summary>
+        void Invalidate(string clientId, string reason);
+
+        /// <summary>
+        /// Clears all entries (test-only; production code should not call this).
+        /// </summary>
+        void ClearForTest();
+    }
+
+    /// <summary>
+    /// Process-wide, identity-scoped, thread-safe certificate cache,
+    /// race-proofed to ensure only one mint/rotation occurs per identity.
+    /// </summary>
+    internal sealed class ManagedIdentityCertificateCache : ICertCache
+    {
+        /// <summary>
+        /// Per-identity container that serializes mint/refresh via a gate.
+        /// </summary>
+        private sealed class Entry
+        {
+            private readonly SemaphoreSlim _gate = new (1, 1);
+            private CertCacheEntry _current;
+
+            internal Entry() { } // First caller will mint behind the gate.
+
+            /// <summary>
+            /// Returns the current binding if still fresh; otherwise mints/rotates exactly once.
+            /// </summary>
+            internal async Task<CertCacheEntry> GetOrRefreshAsync(
+                Func<CancellationToken, Task<CertCacheEntry>> acquireFunc,
+                RequestContext ctx,
+                CancellationToken ct)
+            {
+                var now = DateTimeOffset.UtcNow;
+                if (_current != null && now < _current.RefreshAt)
+                {
+                    return _current;
+                }
+
+                await _gate.WaitAsync(ct).ConfigureAwait(false);
+                try
+                {
+                    now = DateTimeOffset.UtcNow;
+                    if (_current != null && now < _current.RefreshAt)
+                    {
+                        return _current;
+                    }
+
+                    _current = await acquireFunc(ct).ConfigureAwait(false);
+                    ctx.Logger.Info($"[Managed Identity] mTLS cert minted/rotated; expires={_current.Certificate.NotAfter:u}");
+                    return _current;
+                }
+                finally
+                {
+                    _gate.Release();
+                }
+            }
+
+            /// <summary>
+            /// Forces a refresh by marking the current entry as expired.
+            /// </summary>
+            internal void Invalidate()
+            {
+                if (_current != null)
+                {
+                    _current.ForceRefreshNow();
+                }
+            }
+        }
+
+        // Identity map keyed by authoritative client_id (IMDSv2). Case-insensitive for GUID robustness.
+        private readonly ConcurrentDictionary<string, Entry> _entries =
+            new ConcurrentDictionary<string, Entry>(StringComparer.OrdinalIgnoreCase);
+
+        /// <inheritdoc />
+        public async Task<CertCacheEntry> GetAsync(
+            string clientId,
```

Comment:
If you're building a cache, then this is not the right name. "Key" should be used.